lib/ziti_tunnel_cbs.c

```diff
@@ -41,20 +41,41 @@ ssize_t on_ziti_data(ziti_connection conn, uint8_t *data, ssize_t len) {
             ziti_sdk_c_close(ziti_io_ctx);
         }
         return accepted;
-    } else {
-        ZITI_LOG(INFO, "ziti service closed connection");
+    } else if (len == ZITI_EOF) {
+        ZITI_LOG(DEBUG, "ziti connection sent EOF (ziti_eof=%d, tnlr_eof=%d)", ziti_io_ctx->ziti_eof, ziti_io_ctx->tnlr_eof);
+        ziti_io_ctx->ziti_eof = true;
+        if (ziti_io_ctx->tnlr_eof) /* both sides are closed now */ {
+            ziti_tunneler_close(&ziti_io_ctx->tnlr_io_ctx);
+            free(ziti_io_ctx);
+        } else {
+            ziti_tunneler_close_write(&ziti_io_ctx->tnlr_io_ctx);
+        }
+    } else if (len < 0) {
+        ZITI_LOG(ERROR, "ziti connection is closed due to [%zd](%s)", len, ziti_errorstr(len));
         ziti_tunneler_close(&ziti_io_ctx->tnlr_io_ctx);
+        free(ziti_io_ctx);
     }
     return len;
 }
 
 /** called by tunneler SDK after a client connection is closed */
-void ziti_sdk_c_close(void *ziti_io_ctx) {
-    ziti_io_context *_ziti_io_ctx = ziti_io_ctx;
-    if (_ziti_io_ctx->ziti_conn != NULL) {
-        ziti_close(&_ziti_io_ctx->ziti_conn);
+int ziti_sdk_c_close(void *io_ctx) {
```

Comment:
I'd prefer to name this arg `ziti_io_ctx` or similar. io_ctx could mislead someone to pass a struct io_ctx_s *.